### DISCUSSION TITLE
FMU compiled to WebAssembly fails during fmi2ExitInit

### DISCUSSION THREAD

**Discussion by TomasKulhanek** [OpenModelica/OpenModelica#11791] [Q&A]
Dear all,

thanks for updating OM with improved FMU export. 
I'm now pursuing to do FMU export in co-simulation mode with CVODE solver and make a reusable script for compiling it to WebAssembly. 

However, I tried to export a model to FMU with EULER solver or CVODE solver and compiled it to WebAssembly using versions OpenModelica nightly (1.23) as well as 1.22 and 1.20.

Executing both FMUs (EULER and CVODE fmus) webassembly in a browser fails during the fmi2ExitInitializationMode. 

```
FMU(PulsatileCirculationEuler:0:logFmi2Call) msg: fmi2EnterInitializationMode...
VM9440:5356 FMU(PulsatileCirculationEuler:0:logFmi2Call) msg: fmi2ExitInitializationMode...
fmi.js:529 initialize() fmiExitInit throws err RuntimeError: unreachable
    at 005690ba:0x1d81
    at 005690ba:0x114baa
    at 005690ba:0x111165
    at 005690ba:0x110d56
    at 005690ba:0xd160b
    at invoke_iiiiiiiiiii (<anonymous>:5543:32)
    at 005690ba:0xdcc3f
    at invoke_iii (<anonymous>:5345:32)
    at 005690ba:0xd6be0
    at 005690ba:0x10b151
```

I did some more debug logs to localize the problem it seems to fail on the call defined inside source file: `sources/fmi-export/fmu2_model_interface.c.inc`
```
    if (initialization(comp->fmuData, comp->threadData, "fmi", "", 0.0))
``` 

1. For FMU with EULER, no changes in FMU sources were made and I compiled it using the script at: https://github.com/creative-connections/Bodylight.js-FMU-Compiler/blob/master/compiler/openmodelicaeuler.sh
2. For FMU with CVODE, some changes to sources to FMU are necessary in order to compile. ths script is at The script I use is at https://github.com/creative-connections/Bodylight.js-FMU-Compiler/blob/master/compiler/openmodelicacvode.sh
  a. changes to src in order to upgrade from SUNDIALS 5.x to SUNDIALS 6.x as the newer version can be compiled to webassembly easier way as they introduce some cross-compiling capabilities) 
  b. changes to src to read flags in JSON file as in webassembly it is provided in a virtual file system and `fmu_read_flags.c.inc` seems to have wrong `resourceDir`. 
  c. patch for FMU sources I use are in directory https://github.com/creative-connections/Bodylight.js-FMU-Compiler/tree/master/compiler/fmudiff

I would appreciate any help in order to progress with FMU compiling to WebAssembly and executing models in Web Browser. ~~Note that FMUs with EULER solver compiled with our script used to work when exported from OpenModelica 1.18.~~(updated 11/01/2023, issue reproduced using OM 1.19 and OM 1.23)

The issue can be reproduced e.g. at https://egolem.online/omcvodesimulator/#index1.md

Attached FMU with EULER and CVODE:

[PulsatileCirculationEuler.fmu.zip](https://github.com/OpenModelica/OpenModelica/files/13854221/PulsatileCirculationEuler.fmu.zip)

[PulsatileCirculationCvode.fmu.zip](https://github.com/OpenModelica/OpenModelica/files/13854244/PulsatileCirculationCvode.fmu.zip)

**Comment by casella** [comment]
@arun3688 can you please comment on that?

**Comment by TomasKulhanek** [comment]
To update this issue: I tried to export the same model with older version of OpenModelica 1.19 with Euler solver and compiled with latest version of emscripten. It seems to produce similar error in WASM (initialize() fmiExitInit throws err RuntimeError: unreachable). Thus the information that it "used to work with older OM" seems to be misleading. I'll continue to dissect the problem with different emscripten version.

**Comment by TomasKulhanek** [comment]
Trying to reproduce the issue on the older OM shows that the problem is there as well. The complex model 'PulsatileCirculation' comes from Physiolibrary 2.3.1/Hydraulic/Examples/Fernandes2013/PulsatileCirculation.mo. 

The original problem isn't there when exporting simple model (Lorenz Attractor
[LorenzAttractor.zip](https://github.com/OpenModelica/OpenModelica/files/13919729/LorenzAttractor.zip)
) FMU with Euler solver, and it works in Webbrowser. Working demo: https://egolem.online/omcvodesimulator/#lorenzattractor2024.md

But this simple model with CVODE solver compiled with emscripten to Webassembly throws another error after initizalization: `##CVODE## SUNDIALS not available in FMU`, full logs: 
```
FMU(LorenzAttractor2024CVODE2:0:logAll) msg: fmi2Instantiate: Trying to find simulation settings LorenzAttractor2024_flags.json.
FMU(LorenzAttractor2024CVODE2:0:logAll) msg: fmi2Instantiate: Found simulation settings LorenzAttractor2024_flags.json.
FMU(LorenzAttractor2024CVODE2:0:logAll) msg: Initializing CVODE ODE Solver
##CVODE## SUNDIALS not available in FMU. See OpenModelica command line flag "--fmiFlags" from "omc --help" on how to enable CVODE in FMUs.
FMU(LorenzAttractor2024CVODE2:0:logFmi2Call) msg: fmi2Instantiate: GUID={8c4e810f-3df3-4a00-8276-176fa3c9f9e0}
...
FMU(LorenzAttractor2024CVODE2:0:logFmi2Call) msg: fmi2GetContinuousStates: #r2# = 0
##CVODE## SUNDIALS not available in FMU. See OpenModelica command line flag "--fmiFlags" from "omc --help" on how to enable CVODE in FMUs.
FMU(LorenzAttractor2024CVODE2:4:logStatusFatal) msg: fmi2DoStep: CVODE integrator step failed.
```
So next to the original problem above, can you also check, please, compilation scripts https://github.com/creative-connections/Bodylight.js-FMU-Compiler/blob/master/compiler/openmodelicacvode.sh
Web app to reproduce issue - visible in dev. console: https://egolem.online/omcvodesimulator/#lorenzattractor2024cvode.md
